File: merlin-sdk/src/main/java/gov/nasa/jpl/ammos/mpsa/aerie/merlinsdk/activities/Activity.java
package gov.nasa.jpl.ammos.mpsa.aerie.merlinsdk.activities;

import gov.nasa.jpl.ammos.mpsa.aerie.merlinsdk.annotations.ActivityType;

import java.util.List;

/**
 * A mission-specific representation of an activity.
 *
 * Mission activities should implement this interface, as well as the {@link ActivityType}
 * protocol. Implementations of this interface provide methods used by the Merlin system
 * to interact with activity instances.
 */
public interface Activity {
  /**
   * Checks if this activity instance is valid according to mission-specific criteria.
   *
   * @return A list of validation failures, or an empty list if no failures occurred.
   */
  default List<String> validateParameters() { return List.of(); }

  /**
   * Performs the effects of simulating this activity.
   */
  default void modelEffects() { }
}

File: procedural/timeline/src/main/kotlin/gov/nasa/ammos/aerie/procedural/timeline/payloads/activities/Activity.kt
package gov.nasa.ammos.aerie.procedural.timeline.payloads.activities

import gov.nasa.jpl.aerie.merlin.protocol.types.Duration
import gov.nasa.ammos.aerie.procedural.timeline.payloads.IntervalLike

/** Unifying interface for activity instances and directives. */
interface Activity<A: IntervalLike<A>>: IntervalLike<A> {
  /** String type name of the activity. */
  val type: String

  /** Time that the activity starts. */
  val startTime: Duration
}
